[GH-2504] Geopandas: Implement force_3d

chay0112 opened the pull request:
## Did you read the Contributor Guide?

- Yes, I have read the [Contributor Rules](https://sedona.apache.org/latest/community/rule/) and [Contributor Development Guide](https://sedona.apache.org/latest/community/develop/)

## Is this PR related to a ticket?

- Yes, and the PR name follows the format `[GH-XXX] my subject`. Closes #2504 

## What changes were proposed in this PR?

- Implement force_3d

## How was this patch tested?

- Included unit and parity tests

## Did this PR include necessary documentation updates?

- Yes, I have updated the documentation.

petern48 commented on python/sedona/spark/geopandas/base.py:
```diff
@@ -949,8 +949,68 @@ def force_2d(self):
+        Examples
+        --------
+        >>> from shapely import Polygon, LineString, Point
+        >>> s = geopandas.GeoSeries(
```
```suggestion
        >>> from shapely import Polygon, LineString, Point
        >>> from sedona.spark.geopandas import GeoSeries
        >>> s = GeoSeries(
```
nit

petern48 commented on python/tests/geopandas/test_geoseries.py:
```diff
@@ -1520,7 +1520,79 @@ def test_force_2d(self):
+                Point(0.5, 2.5, 2),
+                Point(1, 1, 2),
+                LineString([(1, 1, 2), (0, 1, 2), (1, 0, 2)]),
+                Polygon([(0, 0, 3), (0, 10, 3), (10, 10, 3), (0, 0, 3)]),
```
```suggestion
                LineString([(1, 1, 3), (0, 1, 3), (1, 0, 3)]),
                Polygon([(0, 0, 4), (0, 10, 4), (10, 10, 4), (0, 0, 4)]),
```

I noticed our tests here were slightly buggy. Here's what the expected results should look like. These geoms correspond to the 3 and 4 in `z = [0, 2, 2, 3, 4, 5]`. What's concerning is that the tests pass anyway. Experimenting locally, I noticed we can change the z-coordinate however we like, and the tests still pass. I think what's happening is that the `check_sgpd_equals_gpd()` doesn't yet check anything past the x and y dimensions...

(sigh) Unfortunately, what this means is that we'll need to modify the test code to test this properly before we can merge this PR. I filed an issue for this: https://github.com/apache/sedona/issues/2513. You're welcome to try taking a stab at it or just move on to something else (while keeping this PR open). My time is limited, so I wouldn't be able to investigate it for at least a few days. I don't love having to put a hold on merging your great work here, but this is just how it is sometimes. Sorry about this :(

petern48 commented:
Alright @chay0112, the [fix](https://github.com/apache/sedona/pull/2517) has been merged. Think this should be ready once that last suggestion is accepted.

chay0112 commented:
> Alright @chay0112, the [fix](https://github.com/apache/sedona/pull/2517) has been merged. Think this should be ready once that last suggestion is accepted.

Thanks for your fix Peter, I believe I need to rephrase my tests in test_match_geopandas_series.py because things like Line Strings to Linear Rings and Polygons to Multipolygons would fail

chay0112 commented:
@petern48 I made some adjustments to the tests in test_match_geopandas_series.py because the previous assertions were failing for M and ZM geometry cases. Please review the changes and let me know if there’s a better or cleaner approach. Feedback is welcome.

chay0112 commented:
> Responding to the deletions in this commit: [f90875a](https://github.com/apache/sedona/commit/f90875aae014649e325cbc63ef9de748a6afc739). Here are the failures from the test_match_geopandas tests (I ran locally). Next time, ideally report back when you see behavior like this, instead of relying on me to investigate myself (sometimes, it may take longer that way). Sometimes, it's reasonable to skip; sometimes it's major enough to just put a halt to it. Alright let's see.
> 
> Error for LinearRing:
> 
> ```
> ValueError: Geometry equality check failed for LINESTRING Z (0 0 4, 1 0 4, 1 1 4, 0 1 4, 0 0 4) and LINEARRING Z (0 0 4, 1 0 4, 1 1 4, 0 1 4, 0 0 4)
> ```
> 
> Sedona is returning the equivalent of a LineString. Sedona doesn't support LinearRings since LineStrings are enough. Yep this is fine to skip ✅.
> 
> Error for MultiPolygon:
> 
> ```
> ValueError: Geometry equality check failed for POLYGON Z ((0 0 4, 0 1 4, 1 0 4, 0 0 4), (0.1 0.1 4, 0.1 0.2 4, 0.2 0.1 4, 0.1 0.1 4)) and MULTIPOLYGON Z (((0 0 4, 0 1 4, 1 0 4, 0 0 4), (0.1 0.1 4, 0.1 0.2 4, 0.2 0.1 4, 0.1 0.1 4))
> ```
> 
> Ah interesting, it's the same geometry, but we return polygon instead of multipolygon. This is a minor bug in Sedona we found. Ideally, we want to return MultiPolygon. I dug in real quick and submitted a fix for this: #2526
> 
> Trying it on PostGIS, I see returning MultiPolygon is the desired behavior as well
> 
> ```
> SELECT ST_AsText(ST_Force3D(ST_GeomFromText('MULTIPOLYGON(((0 0, 0 1, 1 0, 0 0), (0.1 0.1, 0.1 0.2, 0.2 0.1, 0.1 0.1)))'), 4));
> -- MULTIPOLYGON Z (((0 0 4,0 1 4,1 0 4,0 0 4),(0.1 0.1 4,0.1 0.2 4,0.2 0.1 4,0.1 0.1 4)))
> ```
> 
> Error for GeometryCollection
> 
> ```
> ValueError: Geometry equality check failed for GEOMETRYCOLLECTION Z (MULTIPOINT Z ((0 0 4), (1 1 4)), MULTILINESTRING Z ((0 0 4, 1 1 4), (2 2 4, 3 3 4)), POLYGON Z ((0 0 4, 0 1 4, 1 0 4, 0 0 4), (0.1 0.1 4, 0.1 0.2 4, 0.2 0.1 4, 0.1 0.1 4))) and GEOMETRYCOLLECTION Z (MULTIPOINT Z ((0 0 4), (1 1 4)), MULTILINESTRING Z ((0 0 4, 1 1 4), (2 2 4, 3 3 4)), MULTIPOLYGON Z (((0 0 4, 0 1 4, 1 0 4, 0 0 4), (0.1 0.1 4, 0.1 0.2 4, 0.2 0.1 4, 0.1 0.1 4))))
> ```
> 
> Fails for the same reason as multipolygon (since it has a multipolygon inside of it).
> 
> Error for M coordinates. Hmm, yeah this one is interesting. Give me a bit more time to figure out what to do here.
> 
> ```
> # For POINT M (1 2 3) input
> ValueError: Geometry equality check failed for POINT Z (1 2 7.5) and POINT M (1 2 3)
> 
> # For POINT ZM (1 2 3 4) input
> ValueError: Geometry equality check failed for POINT Z (1 2 3) and POINT Z (1 2 7.5)
> ```

@petern48 Sure, I'll report the anomalies from next time, Thanks for a very clear explanation. Take your time. Thanks

petern48 commented on python/tests/geopandas/test_geoseries.py:
```diff
@@ -1520,7 +1520,79 @@ def test_force_2d(self):
+            ]
+        )
+        result = s.force_3d(z)
+        self.check_sgpd_equals_gpd(result, expected)
```
For those curious about the difference in behavior, Geopandas returns the following instead. It's strange and inconsistent imo... I would guess it's unintentional and just a result of adding M support later on.

```python
gpd.GeoSeries([shapely.wkt.loads("POINT M (1 2 3)")]).force_3d(7.5)
0    POINT M (1 2 3)

gpd.GeoSeries([shapely.wkt.loads("POINT ZM (1 2 3 4)")]).force_3d(7.5)
0    POINT Z (1 2 7.5)
```

petern48 reviewed (commented):
Sorry for the delay @chay0112. Thanks for being patient. I took another look and saw that Sedona follows PostGIS behavior (which is what we want), and Geopandas behavior apparently is just different. I think we can keep this difference in behavior, but specify this difference in the docstring. Left some suggestions, but I think this should be ready once addressed.

petern48 commented:
Also, could you mention "geopandas" in the PR title (somehow). The PR title eventually becomes the commit message when merged, so it's helpful to specify these details. Could be something like one of the following (I typically do the first one).
- [GH-XXXX] Geopandas: Implement force_3d
- [GH-XXXX] Implement geopandas force_3d

chay0112 commented on python/tests/geopandas/test_match_geopandas_series.py:
```diff
@@ -878,7 +878,48 @@ def test_force_2d(self):
+            pytest.skip("geopandas force_3d requires version 1.0.0 or higher")
+        # 1) Promote 2D to 3D with z = 4
+        for geom in self.geoms:
+            if isinstance(geom[0], (LinearRing, GeometryCollection, MultiPolygon)):
```
Hi Peter, Thanks for your suggestions.
I believe if we remove Multipolygon and GeometryCollections the test would still fail for geoms, something like below
`

FAILED
============================================================================== FAILURES ==============================================================================
_______________________________________________________________ TestMatchGeopandasSeries.test_force_3d _______________________________________________________________

self = <tests.geopandas.test_match_geopandas_series.TestMatchGeopandasSeries object at 0x154827250>

    def test_force_3d(self):
        # force_3d was added from geopandas 1.0.0
        if parse_version(gpd.__version__) < parse_version("1.0.0"):
            pytest.skip("geopandas force_3d requires version 1.0.0 or higher")
        # 1) Promote 2D to 3D with z = 4
        for geom in self.geoms:
            if isinstance(geom[0], (LinearRing)):
                continue
            sgpd_result = GeoSeries(geom).force_3d(4)
            gpd_result = gpd.GeoSeries(geom).force_3d(4)
>           self.check_sgpd_equals_gpd(sgpd_result, gpd_result)

python/tests/geopandas/test_match_geopandas_series.py:890: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
python/tests/geopandas/test_geopandas_base.py:70: in check_sgpd_equals_gpd
    cls.assert_geometry_almost_equal(a, e, tolerance)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

cls = <class 'tests.geopandas.test_match_geopandas_series.TestMatchGeopandasSeries'>
left_geom = <POLYGON Z ((0 0 4, 0 1 4, 1 0 4, 0 0 4), (0.1 0.1 4, 0.1 0.2 4, 0.2 0.1 4, ...>
right_geom = <MULTIPOLYGON Z (((0 0 4, 0 1 4, 1 0 4, 0 0 4), (0.1 0.1 4, 0.1 0.2 4, 0.2 0...>, tolerance = 0.01

    @classmethod
    def assert_geometry_almost_equal(
        cls,
        left_geom: Union[str, BaseGeometry],
        right_geom: Union[str, BaseGeometry],
        tolerance=1e-6,
    ):
        """
        Assert that two geometries are almost equal.
    
        Note: this function will only check Z and M dimensions for shapely >= 2.1.0 (python >= 3.10)
    
        When comparing geometries with Z or M dimensions, this function will ignore `tolerance` and check for exact equality.
        """
        expected_geom = (
            wkt.loads(left_geom) if isinstance(left_geom, str) else left_geom
        )
        actual_geom = (
            wkt.loads(right_geom) if isinstance(right_geom, str) else right_geom
        )
    
        # Note: only shapely >= 2.1.0 supports Z and M dimensions
        # If has Z or M dimension, use equals_identical to check the equality
        if SHAPELY_GE_210 and (has_zm(actual_geom) or has_zm(expected_geom)):
            if not shapely.equals_identical(actual_geom, expected_geom):
>               raise ValueError(
                    f"Geometry equality check failed for {left_geom} and {right_geom}"
                )
E               ValueError: Geometry equality check failed for POLYGON Z ((0 0 4, 0 1 4, 1 0 4, 0 0 4), (0.1 0.1 4, 0.1 0.2 4, 0.2 0.1 4, 0.1 0.1 4)) and MULTIPOLYGON Z (((0 0 4, 0 1 4, 1 0 4, 0 0 4), (0.1 0.1 4, 0.1 0.2 4, 0.2 0.1 4, 0.1 0.1 4)))

python/tests/test_base.py:146: ValueError
========================================================================== warnings summary ==========================================================================
`

chay0112 replied:
Cool, Thanks for letting me know.

chay0112 commented:
@petern48 Just addressed all the comments. It looks like the docs build is failing, and I noticed the same issue in other PRs as well. I'm not sure of the root cause, but if that's not a blocker, this PR is ready for review. No rush though !
